File: apps/mobile/modules/fast-bridge-tls/ios/FastBridgeTlsModule.swift
import CryptoKit
import ExpoModulesCore

private func certificateFingerprint(_ certificate: SecCertificate) -> String {
  let der = SecCertificateCopyData(certificate) as Data
  let digest = SHA256.hash(data: der)
  return "sha256:" + digest.map { String(format: "%02x", $0) }.joined()
}

private func normalizePin(_ raw: String?) -> String? {
  guard var hex = raw?.lowercased() else { return nil }
  if hex.hasPrefix("sha256:") { hex = String(hex.dropFirst(7)) }
  hex = hex.replacingOccurrences(of: "[^0-9a-f]", with: "", options: .regularExpression)
  return hex.isEmpty ? nil : "sha256:\(hex)"
}

private final class ProbeDelegate: NSObject, URLSessionDelegate {
  var fingerprint: String?
  var failure: String?

  func urlSession(
    _ session: URLSession,
    didReceive challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
    defer { completionHandler(.cancelAuthenticationChallenge, nil) }
    guard let trust = challenge.protectionSpace.serverTrust,
      let chain = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
      let leaf = chain.first
    else {
      failure = "no server certificate"
      return
    }
    fingerprint = certificateFingerprint(leaf)
  }
}

private final class SocketDelegate: NSObject, URLSessionDelegate, URLSessionWebSocketDelegate {
  let pin: String
  var onOpen: (() -> Void)?
  var onFail: ((String) -> Void)?
  var onClose: (() -> Void)?

  init(pin: String) {
    self.pin = pin
  }

  func urlSession(
    _ session: URLSession,
    didReceive challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
    guard let trust = challenge.protectionSpace.serverTrust,
      let chain = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
      let leaf = chain.first
    else {
      onFail?("no server certificate")
      completionHandler(.cancelAuthenticationChallenge, nil)
      return
    }
    let got = certificateFingerprint(leaf)
    if got == pin {
      completionHandler(.useCredential, URLCredential(trust: trust))
    } else {
      onFail?("server fingerprint \(got) does not match pinned \(pin)")
      completionHandler(.cancelAuthenticationChallenge, nil)
    }
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    onOpen?()
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
    reason: Data?
  ) {
    onClose?()
  }
}

public class FastBridgeTlsModule: Module {
  private var session: URLSession?
  private var task: URLSessionWebSocketTask?
  private var socketDelegate: SocketDelegate?

  public func definition() -> ModuleDefinition {
    Name("FastBridgeTls")
    Events("open", "message", "error", "close")

    AsyncFunction("probe") { (url: String, expected: String?, promise: Promise) in
      guard let target = URL(string: url) else {
        promise.reject("ERR_INVALID_URL", "invalid url: \(url)")
        return
      }
      let delegate = ProbeDelegate()
      let session = URLSession(configuration: .ephemeral, delegate: delegate, delegateQueue: nil)
      var request = URLRequest(url: target)
      request.timeoutInterval = 10
      request.httpMethod = "GET"
      let task = session.dataTask(with: request) { _, _, _ in
        session.finishTasksAndInvalidate()
        if let fingerprint = delegate.fingerprint {
          if let pin = normalizePin(expected), fingerprint != pin {
            promise.reject(
              "ERR_FINGERPRINT_MISMATCH",
              "server fingerprint \(fingerprint) does not match pinned \(pin)"
            )
          } else {
            promise.resolve(fingerprint)
          }
        } else {
          promise.reject("ERR_PROBE_FAILED", delegate.failure ?? "TLS handshake failed")
        }
      }
      task.resume()
    }

    AsyncFunction("connect") { (url: String, fingerprint: String, promise: Promise) in
      guard let target = URL(string: url), let pin = normalizePin(fingerprint) else {
        promise.reject("ERR_INVALID_URL", "invalid url or fingerprint")
        return
      }
      self.disconnectSocket()
      var settled = false
      let delegate = SocketDelegate(pin: pin)
      delegate.onOpen = {
        if !settled {
          settled = true
          promise.resolve(nil)
        }
        self.sendEvent("open", [:])
        self.receiveLoop()
      }
      delegate.onFail = { message in
        if !settled {
          settled = true
          promise.reject("ERR_CONNECT_FAILED", message)
        }
        self.sendEvent("error", ["message": message])
        self.sendEvent("close", [:])
      }
      delegate.onClose = {
        self.sendEvent("close", [:])
      }
      self.socketDelegate = delegate
      let session = URLSession(configuration: .ephemeral, delegate: delegate, delegateQueue: nil)
      self.session = session
      let task = session.webSocketTask(with: target)
      self.task = task
      task.resume()
    }

    Function("send") { (text: String) -> Bool in
      guard let task = self.task else { return false }
      task.send(.string(text)) { _ in }
      return true
    }

    Function("disconnect") {
      self.disconnectSocket()
    }
  }

  private func receiveLoop() {
    task?.receive { [weak self] result in
      guard let self else { return }
      switch result {
      case .success(.string(let text)):
        self.sendEvent("message", ["data": text])
        self.receiveLoop()
      case .success:
        self.receiveLoop()
      case .failure(let error):
        self.sendEvent("error", ["message": error.localizedDescription])
        self.sendEvent("close", [:])
      }
    }
  }

  private func disconnectSocket() {
    task?.cancel(with: .goingAway, reason: nil)
    session?.invalidateAndCancel()
    task = nil
    session = nil
    socketDelegate = nil
  }
}

File: apps/mobile/modules/fast-bridge-tls/android/src/main/java/expo/modules/fastbridgetls/FastBridgeTlsModule.kt
package expo.modules.fastbridgetls

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSocketFactory
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

class FastBridgeTlsModule : Module() {
  private val executor = Executors.newSingleThreadExecutor()
  private var socket: WebSocket? = null
  private var http: OkHttpClient? = null
  private val generation = AtomicInteger(0)

  override fun definition() = ModuleDefinition {
    Name("FastBridgeTls")
    Events("open", "message", "error", "close")

    AsyncFunction("probe") { url: String, expected: String?, promise: Promise ->
      executor.execute {
        try {
          val fingerprint = probe(url)
          val pin = normalize(expected)
          if (pin != null && fingerprint != pin) {
            promise.reject(
              "ERR_FINGERPRINT_MISMATCH",
              "server fingerprint $fingerprint does not match pinned $pin",
              null
            )
          } else {
            promise.resolve(fingerprint)
          }
        } catch (e: Exception) {
          promise.reject("ERR_PROBE_FAILED", e.message ?: "TLS handshake failed", e)
        }
      }
    }

    AsyncFunction("connect") { url: String, fingerprint: String, promise: Promise ->
      executor.execute {
        try {
          openSocket(url, fingerprint, promise)
        } catch (e: Exception) {
          promise.reject("ERR_CONNECT_FAILED", e.message ?: "无法连接", e)
        }
      }
    }

    Function("send") { text: String ->
      socket?.send(text) == true
    }

    Function("disconnect") {
      generation.incrementAndGet()
      socket?.close(1000, "bye")
      socket = null
      http?.dispatcher?.executorService?.shutdown()
      http = null
    }
  }

  private fun openSocket(url: String, fingerprint: String, promise: Promise) {
    val pin = normalize(fingerprint) ?: throw IllegalArgumentException("invalid fingerprint")
    val gen = generation.incrementAndGet()
    socket?.cancel()
    socket = null
    val trust = pinningTrust(pin)
    val context = SSLContext.getInstance("TLS")
    context.init(null, arrayOf<TrustManager>(trust), SecureRandom())
    val client = OkHttpClient.Builder()
      .sslSocketFactory(context.socketFactory, trust)
      .hostnameVerifier { _, _ -> true }
      .connectTimeout(10, TimeUnit.SECONDS)
      .readTimeout(0, TimeUnit.SECONDS)
      .build()
    http = client
    val request = Request.Builder().url(url).build()
    var settled = false
    socket = client.newWebSocket(
      request,
      object : WebSocketListener() {
        private fun live() = gen == generation.get()

        override fun onOpen(webSocket: WebSocket, response: Response) {
          if (!live()) return
          if (!settled) {
            settled = true
            promise.resolve(null)
          }
          sendEvent("open", emptyMap<String, Any>())
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
          if (live()) sendEvent("message", mapOf("data" to text))
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
          if (!live()) return
          if (!settled) {
            settled = true
            promise.reject("ERR_CONNECT_FAILED", t.message ?: "无法连接", t)
          }
          sendEvent("error", mapOf("message" to (t.message ?: "无法连接")))
          sendEvent("close", emptyMap<String, Any>())
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
          if (live()) sendEvent("close", emptyMap<String, Any>())
        }
      }
    )
  }

  private fun probe(url: String): String {
    val conn = URL(url).openConnection() as HttpsURLConnection
    conn.connectTimeout = 10_000
    conn.readTimeout = 10_000
    conn.sslSocketFactory = trustingSocketFactory()
    conn.hostnameVerifier = HostnameVerifier { _, _ -> true }
    try {
      conn.connect()
      val leaf = conn.serverCertificates.firstOrNull() as? X509Certificate
        ?: throw IllegalStateException("no server certificate")
      return fingerprintOf(leaf)
    } finally {
      conn.disconnect()
    }
  }

  private fun pinningTrust(pin: String): X509TrustManager =
    object : X509TrustManager {
      override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}
      override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
        val leaf = chain.firstOrNull() ?: throw CertificateException("no server certificate")
        val got = fingerprintOf(leaf)
        if (got != pin) throw CertificateException("server fingerprint $got does not match pinned $pin")
      }
      override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
    }

  private fun trustingSocketFactory(): SSLSocketFactory {
    val trustManager = object : X509TrustManager {
      override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}
      override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {}
      override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
    }
    val context = SSLContext.getInstance("TLS")
    context.init(null, arrayOf<TrustManager>(trustManager), SecureRandom())
    return context.socketFactory
  }

  private fun fingerprintOf(cert: X509Certificate): String {
    val digest = MessageDigest.getInstance("SHA-256").digest(cert.encoded)
    return "sha256:" + digest.joinToString("") { "%02x".format(it) }
  }

  private fun normalize(raw: String?): String? {
    if (raw == null) return null
    val hex = raw.lowercase().removePrefix("sha256:").filter { it in '0'..'9' || it in 'a'..'f' }
    return if (hex.isEmpty()) null else "sha256:$hex"
  }
}
